sw/device/tests/usbdev_stream_test.c

```diff
@@ -0,0 +1,753 @@
+
+  CHECK(kDeviceType == kDeviceSimVerilator || kDeviceType == kDeviceFpgaCw310,
+        "This test is not expected to run on platforms other than the "
+        "Verilator simulation or CW310 FPGA. It needs logic on the host side "
+        "to retrieve, scramble and return the generated byte stream");
+
+  LOG_INFO("Running USBDEV Stream Test");
+
+  // Check we can support the requested number of streams
+  CHECK(nstreams && nstreams < USBDEV_NUM_ENDPOINTS);
+
+  // Decide upon the number of bytes to be transferred for the entire test
+  uint32_t transfer_bytes = TRANSFER_BYTES_FPGA;
+  if (kDeviceType == kDeviceSimVerilator) {
+    transfer_bytes = TRANSFER_BYTES_VERILATOR;
+  }
+  transfer_bytes = (transfer_bytes + nstreams - 1) / nstreams;
+  LOG_INFO(" - %u stream(s), 0x%x bytes each", nstreams, transfer_bytes);
+
+  CHECK_DIF_OK(dif_pinmux_init(
+      mmio_region_from_addr(TOP_EARLGREY_PINMUX_AON_BASE_ADDR), &pinmux));
+  pinmux_testutils_init(&pinmux);
+  CHECK_DIF_OK(dif_pinmux_input_select(
+      &pinmux, kTopEarlgreyPinmuxPeripheralInUsbdevSense,
+      kTopEarlgreyPinmuxInselIoc7));
+
+  // Remember context state for usb_testutils context
+  ctx->usbdev = &usbdev;
+
+  // Call `usbdev_init` here so that DPI will not start until the
+  // simulation has finished all of the printing, which takes a while
+  // if `--trace` was passed in.
+  usb_testutils_init(ctx->usbdev, /*pinflip=*/false, /*en_diff_rcvr=*/false,
+                     /*tx_use_d_se0=*/false);
+  usb_testutils_controlep_init(&usbdev_control, ctx->usbdev, 0,
+                               config_descriptors, sizeof(config_descriptors),
+                               test_descriptor, sizeof(test_descriptor));
+  while (usbdev_control.device_state != kUsbTestutilsDeviceConfigured) {
+    usb_testutils_poll(ctx->usbdev);
+  }
+
+  // Initialise the state of each stream
+  for (unsigned id = 0U; id < nstreams; id++) {
+    // Which endpoint are we using for the IN transfers to the host?
+    const uint8_t ep_in = 1u + id;
+    // Which endpoint are we using for the OUT transfers from the host?
+    const uint8_t ep_out = 1u + id;
+    stream_init(ctx, &ctx->streams[id], id, ep_in, ep_out, transfer_bytes);
+  }
+
+  // Decide how many buffers each endpoint may queue up for transmission;
+  // we must ensure that there are buffers available for reception, and we
+  // do not want any endpoint to starve another
+  for (unsigned s = 0U; s < nstreams; s++) {
+    // This is slightly overspending the available buffers, leaving the
+    //   endpoints to vie for the final few buffers, so it's important that
+    //   we limit the total number of buffers across all endpoints too
+    unsigned ep = ctx->streams[s].tx_ep;
+    ctx->tx_bufs_queued[ep] = 0U;
+    ctx->tx_bufs_limit[ep] = (MAX_TX_BUFFERS + nstreams - 1) / nstreams;
+  }
+  ctx->tx_queued_total = 0U;
+
+  if (verbose) {
+    LOG_INFO("Commencing data transfer...");
+  }
+
+  bool bDone = false;
```

Comment:
Following the [style guide](https://staging.opentitan.org/book/doc/contributing/style_guides/c_cpp_coding_style.html#naming)
```suggestion
  bool done = false;
```

Reply:
> Thanks @alees24 for the PR, can some of functions defined in this test be used in other tests? If so I would move them the the `usb_testutils`

I've wondered that myself, and it's a very good question. I still need to extend this test to support the other USB transfer types such as Iscohronous and Interrupt transfers, and it's not yet clear to me whether it will be cleaner to have a single test that exercises all of these transfer types together, or whether the core streaming code can be migrated into lib/testing  alongside usb_testutils_simpleserial.

Since it's still a work in progress, I'd rather keep the code here for now, please; to migrate it runs the risk of complicating later PRs and making the sum total much more involved.

We definitely need a test such as this streaming test in which all transfer types are exercised concurrently, to catch any crosstalk between endpoints/streams.

> Hint: You can use bazel to fix the linting failure in this PR https://staging.opentitan.org/book/doc/contributing/bazel_notes.html?highlight=bazel%20quality#linting-software

Thanks. That's a recent update to this PR and on this occasion I must have neglected to lint first.